src/custom_mode/CustomMode.cpp

```diff
@@ -0,0 +1,295 @@
+/****************************************************************************
+ * Copyright (c) 2023 PX4 Development Team.
+ * SPDX-License-Identifier: BSD-3-Clause
+ ****************************************************************************/
+
+#include "CustomMode.hpp"
+
+#include <Eigen/Core>
+#include <Eigen/Geometry>
+#include <px4_ros2/components/node_with_mode.hpp>
+#include <px4_ros2/utils/geometry.hpp>
+
+// Constants
+static const std::string kModeName = "CustomMode";
+static const bool kEnableDebugOutput = true;
+
+using namespace px4_ros2::literals;
+
+// Constructor for the CustomMode class
+CustomMode::CustomMode(rclcpp::Node& node)
+	: ModeBase(node, kModeName), _node(node)
+{
+	// Initialize the trajectory setpoint and odometry local position
+	_trajectory_setpoint =
+		std::make_shared<px4_ros2::TrajectorySetpointType>(*this);
+	_vehicle_local_position =
+		std::make_shared<px4_ros2::OdometryLocalPosition>(*this);
+
+	// Subscribe to vehicle_land_detected
+	_vehicle_land_detected_sub =_node.create_subscription<px4_msgs::msg::VehicleLandDetected>(
+			"/fmu/out/vehicle_land_detected", rclcpp::QoS(1).best_effort(),
+			std::bind(&CustomMode::vehicleLandDetectedCallback, this,
+				  std::placeholders::_1));
+	// Declare the trajectory type parameter with a default value of "circle"
+	_node.declare_parameter("trajectory_type", "circle");
+	_node.get_parameter("trajectory_type", _trajectory_type);
+}
+
+void CustomMode::vehicleLandDetectedCallback(
+	const px4_msgs::msg::VehicleLandDetected::SharedPtr msg)
+{
+	_land_detected = msg->landed;
+}
+
+// OnActivate function generates waypoints and switches to the Execute state
+void CustomMode::onActivate()
+{
+	generateWaypoints();
+	switchToState(State::Execute);
+}
+// OnDeactivate function logs a message when the mode is deactivated
+void CustomMode::onDeactivate()
+{
+	RCLCPP_INFO(_node.get_logger(), "Deactivating CustomMode");
+}
+
+// UpdateSetpoint function updates the setpoint based on the current state
+void CustomMode::updateSetpoint(float dt_s)
+{
+
+	// Switching between states
+	switch (_state) {
+	// Execute state is the main state where the drone follows the waypoints
+	case State::Execute: {
+
+		// Get the next waypoint
+		Eigen::Vector3f target_position = _waypoints[_waypoint_index];
+
+		// Create a trajectory setpoint message
+		px4_msgs::msg::TrajectorySetpoint setpoint;
+
+		// Construct the trajectory setpoint message
+		setpoint.timestamp = _node.now().nanoseconds() / 1000;
+		setpoint.position = {target_position.x(), target_position.y(),
```

Comment:
considering we're not using the other fields, the **TrajectorySetpoint::update()** method that is already present in the library is sufficient. Then we would not need to use our fork.